Update `operator-sdk`

NicolasT opened the pull request:
This set of patches upgrades the `storage-operator` to `operator-sdk` 0.17.0 (not a fun exercise :sob:)

Likely the biggest change is the code handling leader-election, be advised.

In the end, we end up with a `go.mod` which is somewhat sane, so we may be able to depend on libraries part of the `storage-operator` tree (for now) in other packages.

Closes: #2256.

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.10.1

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v010x

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.11.0

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v011x

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.12.0

This also updates the version of Go used to 1.13.10, as required.

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v012x

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.13.0

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v013x

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.14.1

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v014x

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.15.2

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v015x

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.16.0

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v016x

NicolasT committed:
storage-operator: upgrade to `operator-sdk` 0.17.0

See: https://sdk.operatorframework.io/docs/migration/version-upgrade-guide/#v017x

NicolasT committed:
storage-operator: upgrade skaffolding

This commit updates some of the skaffolded project files with the
files generated using `operator-sdk new ...` using `operator-sdk`
v0.17.0.

Note: because of bff55c2, the 'standard' leader election mechanism/code
is *not* used.

See: https://github.com/scality/metalk8s/commit/bff55c256fff5dff9525f86bd0b6b7dfbc4735b7

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

gdemonet commented on storage-operator/cmd/manager/main.go:
```diff
@@ -116,39 +161,47 @@ func main() {
-	if err := mgr.Start(signals.SetupSignalHandler()); err != nil {
-		log.Error(err, "Manager exited non-zero")
-		os.Exit(1)
+	// The ServiceMonitor is created in the same namespace where the operator is deployed
```
Unfortunately, those ServiceMonitors won't be picked up by our prometheus-operator (since it uses the `release=prometheus-operator` label selector, see #2199)... Looked quickly at their code, not sure we can inject custom labels, the `metrics.CreateServiceMonitors` relies on a custom method on `v1.Service` instance called [`GenerateServiceMonitor`](https://github.com/operator-framework/operator-sdk/blob/master/pkg/metrics/service-monitor.go#L73-L104), which basically copies the Service labels for the ServiceMonitor... And the Service created by `metrics.CreateMetricsService` only has one label: `name=<operator-name>`.

So I guess we'd need to patch the ServiceMonitor(s) after creation to add those labels if we wanted a workaround, and probably should open an issue on Operator SDK's tracker.

NicolasT replied:
Well, `CreateServiceMonitors` takes optional functions to modify the `ServiceMonitor` object before it being created (https://github.com/operator-framework/operator-sdk/blob/0fa4c824583d60fab2f5c0de3275d6a8bde56982/pkg/metrics/service-monitor.go#L37). So, adding this `release=prometheus-operator` label to them should fix this, right?

NicolasT replied:
Now, in general, I'm not that fond of the *operator* creating its own `ServiceMonitor`, `Service` and whatnot objects *at runtime*.

NicolasT replied:
In any case, this should not 'block' this PR:

- The same issue exists with the 'old' code, so metrics aren't captured today.
- We wouldn't be doing anything *useful* with these metrics anyway for now.
- Solution (whilst feasible to implement, cfr. comment above) would better wait for #2199 instead of now patching in the `release=prometheus-operator` label
- I'd rather 'deploy a `ServiceMonitor` with the operator deployment' and get rid of this 'runtime generation'.

gdemonet replied:
Didn't catch the `updaters` argument! So we're fine :)

> I'd rather 'deploy a ServiceMonitor with the operator deployment' and get rid of this 'runtime generation'

Well, I kind of agree in a sense, but also like the idea of considering the Operator as a self-bootstrapping installer... For this specific Operator, it doesn't change much because we deploy it with Salt, so still nothing to do for the user, but for an Operator in the general sense, I understand why this approach is the default.

gdemonet reviewed (commented):
Can't speak much for the go.mod changes, only had a small comment about how this SDK manages metrics and how it doesn't fit with our prometheus-operator deployment today. Probably not related to this PR, so feel free to dismiss :)

I'll let @slaperche-scality give his seal of approval ;)

slaperche-scality commented:
> This set of patches upgrades the `storage-operator` to `operator-sdk` 0.17.0 (not a fun exercise sob)

Seeing how upgrade goes in another project, I'm starting to wonder if this SDK still brings more advantages than pain to us :thinking:

NicolasT committed:
storage-operator: fix Enum validation

In 154f097f9, the `Conditions` field was introduced on a `VolumeStatus`.
However, the OpenAPI validation markers were not properly adjusted, so
we end up with an incorrect CRD schema.

See: 154f097f96ba48c287fb5a273589fdda0aebe4eb

NicolasT committed:
storage-operator: remove `k8s:openapi-gen` marker from `PersistentVolumeTemplateSpec`

This is not a top-level structure for which OpenAPI code needs to be
generated.

NicolasT committed:
ci: update to Fedora 32

Fedora 29 is no longer maintained (and hasn't been for a while).

NicolasT commented:
> > This set of patches upgrades the `storage-operator` to `operator-sdk` 0.17.0 (not a fun exercise sob)
> 
> Seeing how upgrade goes in another project, I'm starting to wonder if this SDK still brings more advantages than pain to us

Agree. Now, these changes are actually somewhat limited, in the end, and I did make the effort of going through every version, which is strictly speaking not required.

Now, for another operator project, I *am* considering to move to Kubebuilder instead... Less 'magic' (but also less features in some regards compared to latest versions of OperatorSDK, like in the testing department).

NicolasT committed:
images: install snapshot version of Python K8s client

The latest release doesn't contain fields for server-side apply
meta-information on CRDs, as supported in K8s 1.17 and generated by
`operator-sdk` 0.17.x. As such, installing a snapshot version for now.

See: https://github.com/kubernetes-client/python/commit/c206745dcd491d64b63260c6b6f46e3a3a81b101

NicolasT commented:
Not fond of https://github.com/scality/metalk8s/pull/2544/commits/3efec35b772388e2c2ceabab611070f46b0124a8 but seems required.

slaperche-scality commented on storage-operator/cmd/manager/main.go:
```diff
@@ -80,14 +83,28 @@ func main() {
+
+	// Set default manager options
+	options := manager.Options{
 		Namespace:               namespace,
```
> Note: because of bff55c2, the 'standard' leader election mechanism/code
is *not* used.

Now that I think about it, this was fixed so we can probably come back to the standard election mechanism.
See https://github.com/operator-framework/operator-sdk/issues/1305

NicolasT replied:
Maybe in some other PR at some point :) FWIW, a Kubebuilder-scaffolded operator uses this mechanism (`LeaderElection` in the manager's `Options`) by default.

NicolasT commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue None.

Goodbye nicolast.
